### PR TITLE
Get integration tests working locally; convert all tests to typescript

### PR DESCRIPTION
**Pull request by rictic**
This is a rebased version of #327 that's also now targeting master.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ const depcheck = require('depcheck');
 const fs = require('fs');
 const glob = require('glob');
 const gulp = require('gulp');
-const mocha = require('gulp-mocha');
+const mocha = require('gulp-spawn-mocha');
```
Comment (justinfagnani) [inline gulpfile.js]: interesting... should we always use this?

Comment (justinfagnani) [inline gulpfile.js]: sort require()s

File: runner/context.ts [hunk 2]
```diff
@@ -134,26 +140,27 @@ export class Context extends events.EventEmitter {
...
+          maybePromise.then(resolve, reject);
         }
       });
     };
 
-    // We execute the handlers _sequentially_. This may be slower, but it gives us
-    // a lighter cognitive load and more obvious logs.
+    // We execute the handlers _sequentially_. This may be slower, but it gives
+    // us a lighter cognitive load and more obvious logs.
+    await Promise.resolve();
```
Comment (justinfagnani) [inline runner/context.ts]: what's this for, are you waiting for the microtask queue to empty? why?

File: test/integration/browser.ts [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+ * @license
+ * Copyright (c) 2014 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+import * as lodash from 'lodash';
+import * as cleankill from 'cleankill';
```
Comment (justinfagnani) [inline test/integration/browser.ts]: sort imports

File: test/unit/cli.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+ * Copyright (c) 2014 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import * as _ from 'lodash';
+import * as chai from 'chai';
```
Comment (justinfagnani) [inline test/unit/cli.ts]: sort imports

File: test/unit/config.ts [hunk 5]
```diff
@@ -7,21 +7,20 @@
...
 
 describe('config', function() {
 
   describe('.merge', function() {
 
     it('avoids modifying the input', function() {
-      var one    = {foo: 1};
-      var two    = {foo: 2};
-      var merged = config.merge(one, two);
+      const one    = <any>{foo: 1};
```
Comment (justinfagnani) [inline test/unit/config.ts]: remove alignment

**Comment by justinfagnani** [comment]
LGTM with some nits

File: gulpfile.js (hunk 1 above)
Comment (rictic) [inline gulpfile.js]: Yes, we should always use gulp-spawn-mocha. gulp-mocha's behavior is undefined if you call it more than once.

File: runner/context.ts (hunk 2 above)
Comment (rictic) [inline runner/context.ts]: This was an artifact of debugging that weird error where `done()` throws by more precisely aping the order of operations of the code before it was converted.

Removed.

File: test/unit/cli.ts (hunk 4 above)
Comment (rictic) [inline test/unit/cli.ts]: done

(is this really worth the time and energy we put into it?)

File: test/integration/browser.ts (hunk 3 above)
Comment (rictic) [inline test/integration/browser.ts]: done.

File: test/unit/config.ts (hunk 5 above)
Comment (rictic) [inline test/unit/config.ts]: done

**Pull request merged**